independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapAppModelFactory.java

```diff
@@ -0,0 +1,522 @@
+                                return new CurationResult((AppModel) in.readObject(), () -> getAppModelResolver());
+                            } else {
+                                debug("Cached deployment classpath has expired for %s", localProject.getAppArtifact());
+                            }
+                        } else {
+                            debug("Unsupported classpath cache format in %s for %s", cachedCpPath,
+                                    localProject.getAppArtifact());
+                        }
+                    } catch (IOException e) {
+                        log.warn("Failed to read deployment classpath cache from " + cachedCpPath + " for "
+                                + localProject.getAppArtifact(), e);
+                    }
+                }
+            }
+
+            AppModelResolver appModelResolver = getAppModelResolver();
+            CurationResult curationResult = new CurationResult(appModelResolver
+                    .resolveManagedModel(localProject.getAppArtifact(), Collections.emptyList(), managingProject), new FixedSuppler<>(appModelResolver));
+            if (cachedCpPath != null) {
+                Files.createDirectories(cachedCpPath.getParent());
+                try (DataOutputStream out = new DataOutputStream(Files.newOutputStream(cachedCpPath))) {
+                    out.writeInt(CP_CACHE_FORMAT_ID);
+                    out.writeInt(localProject.getWorkspace().getId());
+                    ObjectOutputStream obj = new ObjectOutputStream(out);
+                    obj.writeObject(curationResult.getAppModel());
+                } catch (Exception e) {
+                    log.warn("Failed to write classpath cache", e);
+                }
+            }
+            return curationResult;
+        } catch (Exception e) {
+            throw new BootstrapException("Failed to create the application model for " + localProject.getAppArtifact(), e);
+        }
+    }
+
+    /**
+     * If no maven project is around do discovery based on the class path.
+     *
+     * This is used to run gradle tests, and allows them to run from both the IDE
+     * and the gradle test task
+     *
+     */
+    private CurationResult doClasspathDiscovery() {
+        try {
+            AppModelResolver resolver = getAppModelResolver();
+
+            Set<URL> urls = new HashSet<>();
+            //this is pretty yuck, but under JDK11 the URLClassLoader trick does not work
+            Enumeration<URL> manifests = Thread.currentThread().getContextClassLoader().getResources("META-INF/MANIFEST.MF");
+            while (manifests.hasMoreElements()) {
+                URL url = manifests.nextElement();
+                if (url.getProtocol().equals("jar")) {
+                    String path = url.getPath();
+                    if (path.startsWith("file:")) {
+                        path = path.substring(5, path.lastIndexOf('!'));
```

Comment:
The code here is assuming it's a JAR. It could be a directory containing classes, couldn't it?

Reply:
no, see the if statement above.

Reply:
Ok, my point was though about directories containing classes that are on classpath. If you are launching from an IDE that is what should appear on the classpath, right? So why is this method adding only the JARs?

Reply:
I need to think about this one a bit more, but it is really a last resort case when we have no info. I could just delete it but I think it is worth keeping and trying to solve any reported problems as they arise.

Reply:
I think you can keep it for now. AFAIU, this method may get called when (re)building an app from a JAR in the first place. Probably a pretty good chance that everything comes in a JAR at that point.